Create an MLIR environment with matrix multiplication

sogartar opened the pull request:
This also adds config flags to enable the MLIR and LLVM environments. They are mutually exclusive due to their dependence on different LLVM versions and the problem of their coexistence in one CMake configuration.
Enabling them is available through CMake with the flags
```
COMPILER_GYM_ENABLE_MLIR_ENV
COMPILER_GYM_ENABLE_LLVM_ENV
```
The LLVM env is enabled by default.
These flags are propagated to python where sub-modules are included conditionally.

sogartar committed:
Create an MLIR environment with matrix multiplication

Co-authored-by: kyle <kyle.w.herndon@gmail.com>

sogartar commented:
One thing to note is that the CI from the MLIR env is not setup yet.

sogartar committed:
Add to libLLVMPolly.so python package when LLVM env is enabled

sogartar committed:
Fix CMake LLVM version selection

sogartar committed:
Disable MLIR tests in CI

ChrisCummins commented:
Woohoo! Here comes the big PR! 😄 

![](https://media2.giphy.com/media/HKtb2w7S4QQDu/giphy.gif?cid=790b7611751bb73ed3484eaf5351d3ef39b59718c8e2c92b&rid=giphy.gif&ct=g)

There's a lot to unpack here. I'll make a start today, but it might take a couple of rounds for me to go through it all.

Great work @sogartar et al!

Cheers,
Chris

ChrisCummins commented:
For the CI, could you split the existing `CI/build-linux` into two jobs, one which builds the LLVM environment, the other which builds the MLIR environment?

Cheers,
Chris

sogartar committed:
Bump LLVM version for MLIR env to 14.0.0

ChrisCummins commented on INSTALL.md:
```diff
@@ -183,3 +183,10 @@ $ cmake ... \
 
 ⚠️ CompilerGym requires exactly LLVM 10.0.0.
+
+# Contribution
```
This isn't necessary, as `make init` installs the pre-commit hooks:

https://github.com/facebookresearch/CompilerGym/blob/34712e9ec3bd9326bc9f26e025b8c97b74cef8c6/Makefile#L152-L154

sogartar replied:
Done.

ChrisCummins commented on compiler_gym/CMakeLists.txt:
```diff
@@ -49,13 +64,15 @@ cg_py_library(
   PUBLIC
 )
 
+if(COMPILER_GYM_ENABLE_LLVM_ENV)
```
I think actually `compiler_gym::envs::llvm::service::service` is a false dependency. I don't see any reference to it in `random_search.py`. You may be able to remove this data dependency in the CMakeLists.txt/BUILD file.

sogartar replied:
Done.

ChrisCummins commented on compiler_gym/bin/service.py:
```diff
@@ -257,9 +257,7 @@ def print_service_capabilities(env: CompilerEnv):
-            raise NotImplementedError(
-                "Only Commandline and NamedDiscrete are supported."
-            )
+            print(f"{type(action_space)} not supported.")
```
Nitpick:
```suggestion
            print(f"{type(action_space).__name__} not supported.")
```

no?

sogartar replied:
Done.

ChrisCummins commented on compiler_gym/envs/mlir/__init__.py:
```diff
@@ -0,0 +1,60 @@
+        },
+    )
+
+    # TODO(boian): Make better config
+    # observation_spaces = {"runtime": "Runtime"}
+    # reward_spaces = {"runtime": "Runtime"}
+    # for reward_space in reward_spaces:
+    #     register(
+    #         id=f"mlir-{reward_space}-v0",
+    #         entry_point="compiler_gym.envs.mlir:MlirEnv",
+    #         kwargs={
+    #             "service": MLIR_SERVICE_BINARY,
+    #             "reward_space": reward_spaces[reward_space],
+    #         },
+    #     )
+    #
+    # for observation_space, reward_space in product(observation_spaces, reward_spaces):
+    #     register(
+    #         id=f"mlir-{observation_space}-{reward_space}-v0",
+    #         entry_point="compiler_gym.envs.mlir:MlirEnv",
+    #         kwargs={
+    #             "service": MLIR_SERVICE_BINARY,
+    #             "observation_space": observation_spaces[observation_space],
+    #             "reward_space": reward_spaces[reward_space],
+    #         },
+    #     )
```
Should this be committed?

sogartar replied:
@KyleHerndon is working on this, but it is a big change and it will be in another PR.

ChrisCummins commented on compiler_gym/envs/mlir/datasets/__init__.py:
```diff
@@ -0,0 +1,24 @@
+
+
+def get_mlir_datasets(site_data_base: Optional[Path] = None) -> Iterable[Dataset]:
+    """Instantiate the builtin LLVM datasets.
```
```suggestion
    """Instantiate the builtin datasets.
```

sogartar replied:
Done.

ChrisCummins commented on compiler_gym/envs/mlir/datasets/matmul.py:
```diff
@@ -0,0 +1,294 @@
+from compiler_gym.datasets import Benchmark, BenchmarkSource, Dataset
+from compiler_gym.datasets.benchmark import BenchmarkWithSource
+
+# from compiler_gym.envs.mlir.mlir_benchmark import ClangInvocation
```
```suggestion
```

sogartar replied:
Done.

ChrisCummins commented on compiler_gym/envs/mlir/datasets/matmul.py:
```diff
@@ -0,0 +1,294 @@
+
+matmul_sizes = [
+    (64, 64, 64)
+    # (128, 128, 128),
+    # (1024, 1024, 1024)
```
Is this intentional?

sogartar replied:
I removed them. We have improvements in the configuration side in an upcoming PR.

ChrisCummins replied:
Sounds good :)

ChrisCummins commented on compiler_gym/envs/mlir/datasets/matmul.py:
```diff
@@ -0,0 +1,294 @@
+            name="generator://matmul-v0",
+            description="Targeted size matmul programs",
+            references={},
+            license="donotsubmit",
```
Hmm 🙂

sogartar replied:
Fixed to `MIT`.

ChrisCummins commented on compiler_gym/envs/mlir/datasets/matmul.py:
```diff
@@ -0,0 +1,294 @@
+        )
+        self.matmul_bin_path = matmul_bin or _matmul_BIN
+        self.matmul_includes_path = matmul_includes or _matmul_INCLUDES
+        """# The command that is used to compile an mlir-IR bitcode file from a
+        # matmul input. Reads from stdin, writes to stdout.
+        self.clang_compile_command: List[str] = ClangInvocation.from_c_file(
+            "-",  # Read from stdin.
+            copt=[
+                "-xc",  # The C programming language.
+                "-ferror-limit=1",  # Stop on first error.
+                "-w",  # No warnings.
+                f"-I{self.matmul_includes_path}",  # Include the matmul headers.
+            ],
+        ).command(
+            outpath="-"  # Write to stdout.
+        )"""
```
Should this be commented out?

sogartar replied:
I removed it.

ChrisCummins commented on compiler_gym/envs/mlir/datasets/matmul.py:
```diff
@@ -0,0 +1,294 @@
+
+    @property
+    def size(self) -> int:
+        # Actually 2^32 - 1, but practically infinite for all intents and
+        # purposes.
```
```suggestion
```
Comment doesn't seem relevant

sogartar replied:
Done.

ChrisCummins commented on compiler_gym/envs/mlir/datasets/matmul.py:
```diff
@@ -0,0 +1,294 @@
+    def benchmark_uris(self) -> Iterable[str]:
+        return (self.name_from_size(mnk) for mnk in matmul_sizes)
+
+    def benchmark(self, uri: str) -> MatmulBenchmark:
```
I believe you could use [benchmark_from_parsed_uri()](https://compilergym.com/compiler_gym/datasets.html#compiler_gym.datasets.Dataset.benchmark_from_parsed_uri) instead of `benchmark()` to save you the string parsing here

sogartar replied:
This part is going to be changed in another PR.

ChrisCummins commented on compiler_gym/envs/mlir/datasets/matmul.py:
```diff
@@ -0,0 +1,294 @@
+        mnk = list(mnk)
+        # Run matmul with the given size and regex to produce the correct mlir
+        logger.debug("Exec matmul --mnk %d", mnk)
+        """matmul = subprocess.Popen(
```
Not sure if this should be committed

sogartar replied:
Removed.

ChrisCummins commented on compiler_gym/envs/mlir/datasets/mlir_examples/main.cc:
```diff
@@ -0,0 +1,39 @@
+    }
+  }
+}
+\n extern "C" {
```
Is `\n` needed here?

sogartar replied:
I removed it.

ChrisCummins commented:
Hey,

Sorry for the delay in the full review, I just moved to a new office and am setting up a new dev environment. I'm hitting this error, not sure if it's related to your changes or my own fault:

```
$ cmake -DCMAKE_C_COMPILER=clang-7 -DCMAKE_CXX_COMPILER=clang++-7 -DPython3_FIND_VIRTUALENV=FIRST -DCMAKE_BUILD_WITH_INSTALL_RPATH=true -DCOMPILER_GYM_ENABLE_MLIR_ENV=on -DCOMPILER_GYM_ENABLE_LLVM_ENV=off -S $HOME/src/CompilerGym -B /dev/shm/CompilerGymBuilds/mlir


... <snip>

[3/9] Performing build_protobuf step for 'protobuf'

... <snip>

checking whether the C compiler works... no
configure: error: in `/dev/shm/CompilerGymBuilds/mlir/external/protobuf/protobuf/src/protobuf':
configure: error: C compiler cannot create executables
See `config.log' for more details
CMake Error at /private/home/cummins/src/CompilerGym/external/protobuf/build_protobuf.cmake:50 (execute_process):
  execute_process failed command indexes:

    1: "Child return code: 77"

[7/9] No install step for 'protobuf'
FAILED: protobuf/src/protobuf-stamp/protobuf-build_protobuf 
cd /dev/shm/CompilerGymBuilds/mlir/external/protobuf/protobuf/src/protobuf && /private/home/cummins/.local/bin/cmake -E env CC=/bin/clang-7 CXX=/bin/clang++-7 /private/home/cummins/.local/bin/cmake -C /dev/shm/CompilerGymBuilds/mlir/external/protobuf/protobuf_initial_cache.cmake -DGIT_EXECUTABLE=/bin/git -DGIT_REPOSITORY_DIR=/dev/shm/CompilerGymBuilds/mlir/external/protobuf/protobuf/src/protobuf -DCMAKE_INSTALL_PREFIX=/dev/shm/CompilerGymBuilds/mlir/external/protobuf/install -P /private/home/cummins/src/CompilerGym/external/protobuf/build_protobuf.cmake
ninja: build stopped: subcommand failed.
CMake Error at build_tools/cmake/build_external_cmake_project.cmake:37 (execute_process):
  execute_process failed command indexes:

    1: "Child return code: 1"
```

Have you seen this before?

I started reading through the code and left a bunch of minor comments, but I think I'll hold off until I get a chance to build and test the new environment before I go any further. I don't want to spam you with nitpicks until I understand the big picture 🙂 

Note to self: I got as far as `compiler_gym/envs/mlir/mlir_env.py` when reading through the sources

Cheers,
Chris

sogartar committed:
Include clang, MLIRExecutionEngine and Google benchmark python package

sogartar committed:
Add building and testing of MLIR env to CI

sogartar commented:
@ChrisCummins, I added the external dependencies of the MLIR env to the python packaging. It required the clang compiler, Google benchmark and 
MLIRExecutionEngine libraries. I am not sure if this is the best approach or to leave them as runtime dependencies for the user to install.

sogartar committed:
Fix auto-formatting for some files

sogartar committed:
Make CMake to always echo commands to stdout that are ran with execute_process

sogartar committed:
Disable loop_optimizations_service examples when LLVM env is disabled

sogartar committed:
Remove pre-commit instructions from Install.md

sogartar committed:
Remove data dependency of random_search to llvm::service

sogartar committed:
In print_service_capabilities print type name of unspupported action spaces

sogartar committed:
Remove commented out code and fix comments

sogartar commented:
Regarding the failed protobuf configuration. This looks like a problem with the system/compiler. Can you share the protobuf configure log also?
I tried to compile with `clang-7` and on my setup it works fine. I am on `Ubuntu 20.04` and I used `clang-7` from the official deb repo.
I made CMake to print commands ran with [execute_process](https://cmake.org/cmake/help/latest/command/execute_process.html), so you can take the exact command and play with it directly on the protobuf source.

ChrisCummins commented:
> Regarding the failed protobuf configuration. This looks like a problem with the system/compiler.

Yep, I just reproduced the same error on `development`. Sorry for the noise, let me debug.

Cheers,
Chris

sogartar committed:
Add clang include files to MLIR env package and add libstdc++ as a runtime dependence

sogartar committed:
In CI install libstdc++ as root

sogartar committed:
Add ray to python test dependencies

sogartar committed:
Fix dependency to Ray's rllib and tune

sogartar committed:
Shutdown Ray before initializing it in test_ppo_train_smoke

sogartar committed:
In MLIR test_ppo_train_smoke make wrapper env compilant with Gym

sogartar committed:
In MLIR's test_ppo_train_smoke move Ray init before registering the env

sogartar committed:
Increase MLIR compilation timeout from 60 to 120 seconds

sogartar committed:
In MLIR's test_ppo_train_smoke disable deprication warning comming from Ray

sogartar committed:
Ignore more warning in test_ppo_train_smoke

ChrisCummins commented:
Hey guys, I got the build working on a different machine.

I played around with the environment a little and it seems to work, though a little light on features 🙂  The `Runtime` observation seems to always return a single 0 value (ignoring `env.runtime_observation_count`), and the only reward space is `IsRunnable`. Does that look right to you?

```py
In [1]: import compiler_gym

In [2]: compiler_gym.COMPILER_GYM_ENVS
Out[2]: ['gcc-v0', 'mlir-v0', 'loop_tool-v0']

In [3]: env = compiler_gym.make("mlir-v0")

...

In [9]: env.runtime_observation_count = 10

In [10]: env.observation.Runtime()
Out[10]: array([0.])

...

In [14]: env.action_space.sample()
Out[14]:
[OrderedDict([('tile_options',
               OrderedDict([('interchange_vector', array([0, 2, 1])),
                            ('loop_type', 1),
                            ('promote', True),
                            ('promote_full_tile', True),
                            ('tile_sizes',
                             array([ 904669694, 1779810269, 3305806311]))])),
              ('vectorize_options',
               OrderedDict([('unroll_vector_transfers', True),
                            ('vector_transfer_split', 2),
                            ('vectorize_to', 1)]))]),
 OrderedDict([('tile_options',
               OrderedDict([('interchange_vector', array([2, 1, 0])),
                            ('loop_type', 1),
                            ('promote', True),
                            ('promote_full_tile', True),
                            ('tile_sizes',
                             array([1905829399,  245767300, 3150522590]))])),
              ('vectorize_options',
               OrderedDict([('unroll_vector_transfers', True),
                            ('vector_transfer_split', 1),
                            ('vectorize_to', 0)]))]),
 OrderedDict([('tile_options',
               OrderedDict([('interchange_vector', array([2, 1, 0])),
                            ('loop_type', 1),
                            ('promote', True),
                            ('promote_full_tile', True),
                            ('tile_sizes',
                             array([ 908340389, 1436771577, 2394713207]))])),
              ('vectorize_options',
               OrderedDict([('unroll_vector_transfers', True),
                            ('vector_transfer_split', 1),
                            ('vectorize_to', 0)]))])]

...

In [17]: env.reward.spaces
Out[17]: {'IsRunnable': IsRunnable}
```

Cheers,
Chris

sogartar commented:
> I played around with the environment a little and it seems to work, though a little light on features  The `Runtime` observation seems to always return a single 0 value (ignoring `env.runtime_observation_count`), and the only reward space is `IsRunnable`. Does that look right to you?

Before making a step the environment does not compile and run anything. It will return 0. The current environment is actually doing only one step per episode, where all parameters are passed in a single action. It does not run multiple steps where passes are ran on the compiled code from the previous step.
The only reward available should be the running time. I will check why `IsRunnable` is returned.
It is likely that we will upgrade the environment to do more.
I may be wrong but changing `env.runtime_observation_count` would change the number of samples in the observation in a subsequent step or reset.

sogartar committed:
Move RuntimeReward to compiler_gym.spaces

sogartar commented:
I moved the runtime reward from `MlirRlWrapperEnv` to `MlirEnv`. It did not make sense to have it there, since the wrapper is concerned with converting the environment to be interoperable with RL training libraries. 
I decided to move `RuntimeReward` to `compiler_gym.spaces`, so that LLVM and MLIR can both use it. It made most sense to put it there. The problem is that it had a dependency on `BenchmarkInitError` and  `ServiceError`. This introduced a circular dependency. I changed it to throw `ValueError` instead. @ChrisCummins if you think this is an unacceptable change in the interface of `RuntimePointEstimateReward` I can put `RuntimeReward` in the `compiler_gym.envs` package. I don't think this is quite correct because the `envs` package is mostly concerned with holding the available environments.
Another option is to take `RuntimePointEstimateReward` out of the LLVM domain and make it a general wrapper. This will introduce a slight boilerplate when using the `MlirEnv` as it would not have a natural/default reward space.

ChrisCummins commented:
Sounds good! I'll take a look.

I think putting `RuntimeReward` in `compiler_gym.spaces` makes sense.

To break the circular dependency with the custom exception types, we could create a `compiler_gym.errors` module and move the custom exceptions to there. This errors module would have no dependencies. I think there would be some advantage to readability by doing this. There aren't a huge number of custom exception types, but they're pretty spread out. E.g. these four:

* compiler_gym.ValidationError
* compiler_gym.datasets.BenchmarkInitError
* copmiler_gym.service.EnvironmentNotSupported
* compiler_gym.util.DownloadFailed

Coalescing them into a single module would enable a single page on the sphinx HTML docs to list the whole collection.

To maintain backwards compatibility, we could add aliases in the places where custom exceptions are currently defined. E.g. in `compiler_gym/util/download.py`:
```py
# NOTE: Deprecated type alias for backwards compatibility that will be removed in a future release.
DownloadFailed = compiler_gym.errors.DownloadFailed
```

WDYT?

Cheers,
Chris

sogartar commented:
This is a good solution. I will refactor this PR to include this change.

sogartar committed:
Move most Exceptions to compiler_gym.errors

sogartar committed:
Add compiler_gym.errors to setup.py

sogartar commented:
@ChrisCummins, just a reminder that this PR is awaiting another review round.

ChrisCummins commented on compiler_gym/datasets/benchmark.py:
```diff
@@ -312,8 +313,10 @@ def __le__(self, other: Union[str, "Benchmark"]):
-class BenchmarkInitError(OSError):
-    """Base class for errors raised if a benchmark fails to initialize."""
+# Deprecated since v0.2.4.
+# This type is form backwards compatibility that will be removed in a future release.
```
```suggestion
# This type is for backwards compatibility that will be removed in a future release.
```

Tiny typo (repeated elsewhere)

sogartar replied:
Done.

ChrisCummins commented:
> @ChrisCummins, just a reminder that this PR is awaiting another review round.

Great! Taking a look now 🙂 

Thanks for doing the `compiler_gym.errors` refactor!

Cheers,
Chris

ChrisCummins commented on .github/workflows/ci.yaml:
```diff
@@ -92,6 +92,62 @@ jobs:
+                  else
+                    wget https://github.com/llvm/llvm-project/releases/download/llvmorg-14.0.0/clang+llvm-14.0.0-x86_64-linux-gnu-ubuntu-18.04.tar.xz -O ~/llvm.tar.xz
+                  fi
+                  # TODO(cummins): Remove 'v' debugging flag:
```
```suggestion
```

Looks like redundant comment

sogartar replied:
Done.

ChrisCummins commented on .github/workflows/ci.yaml:
```diff
@@ -92,6 +92,62 @@ jobs:
+            - name: Install build dependencies
+              uses: ./.github/actions/install-cmake-build-dependencies
+
+            - name: Download LLVM 10.0.0 release
```
```suggestion
            - name: Download LLVM 14.0.0 release
```

sogartar replied:
Done.

ChrisCummins commented on tests/mlir/mlir_env_test.py:
```diff
@@ -0,0 +1,307 @@
+    assert observation[0] == 0
+
+
+@flaky(max_runs=3, min_passes=1)
+@pytest.mark.filterwarnings(
+    "ignore:`np\\.bool` is a deprecated alias for the builtin `bool`\\.",
+    "ignore:Mean of empty slice",
+    "ignore::ResourceWarning",
+    "ignore:using `dtype=` in comparisons is only useful for `dtype=object`",
+)
+def test_ppo_train_smoke():
+    ray.shutdown()
+    seed = 123
+    np.random.seed(seed)
+    random.seed(seed)
+    torch.manual_seed(seed)
+    ray.init(local_mode=True)  # Runs PPO training in the same process
+    register_env(
+        "mlir_rl_env-v0", lambda env_config: MlirRlWrapperEnv(env=gym.make("mlir-v0"))
+    )
+    config = {
+        "env": "mlir_rl_env-v0",
+        "framework": "torch",
+        # Tweak the default model provided automatically by RLlib,
+        # given the environment's observation- and action spaces.
+        "model": {
+            "fcnet_hiddens": [2, 2],
+            "fcnet_activation": "relu",
+        },
+        "num_workers": 0,  # local worker only
+        "train_batch_size": 2,
+        "sgd_minibatch_size": 1,
+        "num_sgd_iter": 1,
+        "rollout_fragment_length": 2,
+    }
+    trainer = PPOTrainer(config=config)
+    trainer.train()
+    ray.shutdown()
```
IMO This feels a little heavy duty for a unit test. Could you put it in its own file like `tests/mlir/rllib_ppo_smoke_test.py`?

sogartar replied:
Done.

ChrisCummins commented on compiler_gym/bin/service.py:
```diff
@@ -257,9 +257,7 @@ def print_service_capabilities(env: CompilerEnv):
-            raise NotImplementedError(
-                "Only Commandline and NamedDiscrete are supported."
-            )
+            print(f"{type(action_space).__name__} not supported.")
```
How about just printing `print(action_space)` and relying on whatever `__repr__` it defines?

sogartar replied:
Done.

ChrisCummins commented on compiler_gym/envs/mlir/__init__.py:
```diff
@@ -0,0 +1,36 @@
+__all__ = [
+    "MLIR_SERVICE_BINARY",
+    "MlirEnv",
+    "observation_spaces",
+    "reward_spaces",
```
I don't see where these are defined

sogartar replied:
I removed them.

ChrisCummins commented on compiler_gym/envs/mlir/datasets/matmul.py:
```diff
@@ -0,0 +1,268 @@
+logger = logging.getLogger(__name__)
+
+# The maximum value for the --seed argument to matmul.
+UINT_MAX = (2**32) - 1
+
+_matmul_BIN = runfiles_path("compiler_gym/third_party/matmul/matmul/bin/matmul")
+_matmul_INCLUDES = runfiles_path(
+    "compiler_gym/third_party/matmul/matmul/include/matmul-2.3.0"
+)
+
+matmul_sizes = [(64, 64, 64)]
```
Inconsistent mix of `_` prefix and upper/lower case for variable names

sogartar replied:
Done.

ChrisCummins commented on compiler_gym/envs/mlir/service/Benchmark.cc:
```diff
@@ -0,0 +1,443 @@
+  return Status::OK;
+}
+
+// bool Benchmark::applyBaselineOptimizations(unsigned optLevel, unsigned sizeLevel) {
+//   return applyBaselineOptimizationsToModule(&module(), optLevel, sizeLevel);
+// }
```
```suggestion
```

sogartar replied:
Done.

ChrisCummins commented on compiler_gym/envs/mlir/service/Benchmark.h:
```diff
@@ -0,0 +1,328 @@
+    return scratchDirectory_.parent_path();
+  }
+
+  // NOTE(cummins): Order here is important! The LLVMContext must be declared
```
```suggestion
  // NOTE(cummins): Order here is important! The MLIRContext must be declared
```

sogartar replied:
Done.

ChrisCummins commented on compiler_gym/envs/mlir/service/Benchmark.h:
```diff
@@ -0,0 +1,328 @@
+  const boost::filesystem::path scratchDirectory_;
+  const BenchmarkDynamicConfig dynamicConfigProto_;
+  const RealizedBenchmarkDynamicConfig dynamicConfig_;
+  /** The directory used for storing build / runtime artifacts. The difference
+   * between the scratch directory and the working directory is that the working
+   * directory may be shared across multiple Benchmark instances. The scratch
+   * directory is unique.
+   */
```
Comment in the wrong place

sogartar replied:
I moved the comment.

ChrisCummins commented on compiler_gym/envs/mlir/service/CMakeLists.txt:
```diff
@@ -0,0 +1,379 @@
+cg_add_all_subdirs()
+
+set(_DEPS "compiler_gym-mlir-service")
+if(DARWIN)
+    #TODO(boian): figure out what is this target.
+    #list(APPEND _DEPS "@llvm//:darwin")
+endif()
```
Is this needed?

sogartar replied:
Removed.

ChrisCummins commented on compiler_gym/envs/mlir/service/MlirUtils.cc:
```diff
@@ -0,0 +1,285 @@
+}
+
+void LinalgCodegenPass::runOnOperation() {
+  // TODO(kyleherndon): figure out why this won't compile/remove it
```
Should this be done before merging?

sogartar replied:
These parameters are now part of the input MLIR file. I removed this.

ChrisCummins commented on compiler_gym/envs/mlir/service/MlirUtils.h:
```diff
@@ -0,0 +1,30 @@
+
+class ModuleOp;
+
+// TODO(kyleherndon): docstrings
```
🙂

sogartar replied:
I am leaving this to be maybe done at a latter stage. It is pretty deep to have a user facing documentation.

ChrisCummins commented on compiler_gym/service/proto/Proto.cc:
```diff
@@ -0,0 +1,516 @@
+
+#define CG_PROTO_CHECK(cond) \
+  do {                       \
+    if (errorOnFalse) {      \
```
I think `errorOnFalse` should be a macro argument

sogartar replied:
Done.

ChrisCummins commented on compiler_gym/service/proto/Proto.h:
```diff
@@ -0,0 +1,33 @@
+
+namespace compiler_gym {
+
+struct SpaceContainsEventChecker {
```
This needs docstrings

sogartar replied:
Done.

sogartar replied:
I added also a `proto.rst` file to the documentation.

ChrisCummins replied:
Thanks!

ChrisCummins commented on compiler_gym/wrappers/core.py:
```diff
@@ -371,3 +376,70 @@ def multistep(
         raise NotImplementedError
+
+
+class ConversionWrapperEnv(CompilerEnvWrapper):
```
I'm not sure about this. I like that the wrappers in base gym operate only on action/observation/reward at a time. Could you add the space conversion methods to the relevant wrapper and use them instead? E.g. merge `convert_observation_space` into `ObservationWrapper`. You could then have:

```py
env = MlirObservationWrapper(MlirActionWrapper(MlirRewardWrapper(env)))
```

sogartar replied:
Done.

ChrisCummins commented on setup.py:
```diff
@@ -178,6 +174,49 @@ def wheel_filename(**kwargs):
+    setup_kwargs["package_data"]["compiler_gym"].extend(
+        ["envs/mlir/service/compiler_gym-mlir-service"]
+    )
+    origina_cwd = os.getcwd()
```
```suggestion
    original_cwd = os.getcwd()
```

sogartar replied:
Done.

ChrisCummins commented:
Okay, I have completed a full pass of the PR. Minor nitpicks left inline, but in general, this looks excellent. Many thanks @sogartar  and @KyleHerndon!

While this isn't merge blocking, I'd like to discuss with you folks your goals / timeline is for making the MLIR environment a part of the "core" installable package. In particular, I want to avoid this kind of thing:

```py
if config.enable_llvm_env:
  ...

if config.enable_mlir_env:
  ...
```

As I understand, it would be totally possible to build a combined LLVM + MLIR environment package, it would just require a couple of passes of the cmake build and merging the generated runfiles, right? Is the idea to split the cmake build into separate projects for each environment so that they can have mutually exclusive dependencies?

If for some reason there are unavoidable extra dependencies that MLIR requires that complicates the build, could we not still have the python class definition available, and defer the `config.enable_mlir_env` check until instantiation, raising some kind of runtime error like:

```py
>>> compiler_gym.make("mlir-v0")
EnvironmentNotSupported("Please install compiler_gym[mlir] to use this environment")
```

Again, this doesn't need to be resolved in this PR, assuming that you're okay with the pip wheel continuing to be LLVM-only for the time being.

One final request that *is* blocking is for you to add a new page in the documentation describing this environment. Take a look at [envs/gcc](https://github.com/facebookresearch/CompilerGym/blob/development/docs/source/envs/gcc.rst) for an example. Since it is still early days, it would be fine to say that its under construction and that the doc can be updated as more features are added, but it's important to have the basic there.

Cheers,
Chris

sogartar committed:
Fix typos and comments

sogartar committed:
Move MLIR env PPO smoke test to separate file

sogartar committed:
In print_service_capabilities print the string representation of unsupported action spaces

sogartar committed:
Remove some misplaced names for the public interface of compiler_gym.envs.mlir

sogartar committed:
Make errorOnFalse argument of CG_PROTO_CHECK macro

sogartar committed:
Add docstring for class SpaceContainsEventChecker

sogartar committed:
Remove ConversionWrapperEnv

sogartar committed:
Document and add MlirEnv to documatation

sogartar committed:
Fix casing of variables

sogartar committed:
Add proto.rst to documentation

sogartar commented:
> One final request that _is_ blocking is for you to add a new page in the documentation describing this environment. Take a look at [envs/gcc](https://github.com/facebookresearch/CompilerGym/blob/development/docs/source/envs/gcc.rst) for an example.

I added the `MlirEnv` to the documentation. 

> While this isn't merge blocking, I'd like to discuss with you folks your goals / timeline is for making the MLIR environment a part of the "core" installable package.

I am also not happy with the logic of the conditional inclusion of the LLVM and MLIR environments in the package. It needs refactoring. Having multiple subpackages that can be separately built and installed with pip seems a good idea.

I think it would be beneficial to have also a development package that has all stuff related to creating new environments. Like the C++ code in `compiler_gym/service`.

The packages could look like that
```
compiler_gym # runtime
compiler_gym[dev] # development
compiler_gym[gcc]
compiler_gym[llvm]
compiler_gym[loop_tool]
compiler_gym[mlir]
```

@ChrisCummins, I will let you know in a few days when we are going to be able to work on that.

sogartar committed:
Make MLIR test produce test code coverage

ChrisCummins commented:
Thanks for the latest iteration of changes, it's coming together great.

> I added the `MlirEnv` to the documentation.

Ah, what I meant was a new rst doc in `docs/source/envs` which describes in plain english what the optimization problem is and what the features of the environment are. You could start by using [the gcc one](https://github.com/facebookresearch/CompilerGym/blob/development/docs/source/envs/gcc.rst) or [the llvm one](https://github.com/facebookresearch/CompilerGym/blob/development/docs/source/envs/llvm.rst) one as a template.

> I think it would be beneficial to have also a development package that has all stuff related to creating new environments. Like the C++ code in `compiler_gym/service`.

Yep, I'll continue discussion in #669.

Cheers,
Chris

sogartar committed:
Add more extensive documentation for the MLIR env

sogartar commented:
> Ah, what I meant was a new rst doc in `docs/source/envs` which describes in plain english what the optimization problem is and what the features of the environment are. You could start by using [the gcc one](https://github.com/facebookresearch/CompilerGym/blob/development/docs/source/envs/gcc.rst) or [the llvm one](https://github.com/facebookresearch/CompilerGym/blob/development/docs/source/envs/llvm.rst) one as a template.

I added some documentation under `docs/source/envs/mlir.rst`.

sogartar commented:
@ChrisCummins, did you got a chance to see the documentation?

ChrisCummins reviewed (approved):
LGTM, thanks! This looks ready to merge
